### PR TITLE
API for: defining view and aggregations, associating them with metrics, and retrieving data

### PR DESCRIPTION
**Pull request by a-veitch**
Data is returned from all aggregations as a binary blob. The type of data is transparent to the API (so we can define/redefine these with some amount of flexibility).

Also did some renaming to bring in metric nomenclature.

### CODE REVIEW

File: include/grpc/census.h [hunk 1]
```diff
@@ -159,16 +159,154 @@ int census_tag_set_next(census_tag_set_iterator *it, census_tag_const *tag);
...
+   Each metric can have an arbitrary number of views by which it will be
+   broken down. For every measurement recorded, they are broken down by
+   unique tag combinations, and recorded in each matvhing view/aggregation.
+*/
+
+/* A single value to be recorded comprises two parts: an ID for the particular
+ * metric and the value to be recorded against it. */
 typedef struct {
-  int id;
+  gpr_int32 metric_id;
```
Comment (maxwell-demon) [inline include/grpc/census.h]: gpr_uint32?

File: include/grpc/census.h [hunk 2]
```diff
@@ -379,16 +379,131 @@ int census_tag_set_next(census_tag_set_iterator *it, census_tag_const *tag);
...
+void census_record_values(census_context *context, census_value *values,
+                          size_t nvalues);
+
+/** Structure used to describe an aggregation type. */
+typedef struct {
+  /* Create a new aggregation. The pointer returned can be used in future calls
+     to clone(), free(), record(), data() and reset(). */
+  void *(*create)(const void *create_arg);
+  /* Make a copy of an aggregation created by create() */
+  void *(*clone)(const void *aggregation);
```
Comment (maxwell-demon) [inline include/grpc/census.h]: Do we have a user case where cloning an aggregation type is needed?

File: include/grpc/census.h [hunk 3]
```diff
@@ -379,16 +379,131 @@ int census_tag_set_next(census_tag_set_iterator *it, census_tag_const *tag);
...
+  /* Destroy an aggregation created by create() */
+  void (*free)(void *aggregation);
```
Comment (maxwell-demon) [inline include/grpc/census.h]: Maybe just name the function destroy? free is special in C.

File: include/grpc/census.h [hunk 4]
```diff
@@ -379,16 +379,131 @@ int census_tag_set_next(census_tag_set_iterator *it, census_tag_const *tag);
...
+  /* Merge 'from' aggregation into 'to'. Both aggregations must be compatible */
+  void (*merge)(void *to, const void *from);
+  /* Fill buffer with printable string version of aggregation contents. For
+     debugging only. Returns the number of bytes added to buffer (a value == n
+     implies the buffer was of insufficient size). */
+  size_t (*print)(const void *aggregation, char *buffer, size_t n);
+} census_aggregation_ops;
+
+/* Predefined aggregation types. */
+extern census_aggregation_ops census_agg_scalar;
```
Comment (maxwell-demon) [inline include/grpc/census.h]: Maybe call it census_agg_sum or census_agg_total?
min, max, sum, and average etc. are all scalar.

File: include/grpc/census.h (hunk 2 above)
Comment (a-veitch) [inline include/grpc/census.h]: Yes, it is needed internally when we want to create a new aggregation. We either have to clone/zero an existing one, or keep a copy of the initialization argument (which implies that we need more functions in census_aggregation_ops to clone and free that initialization argument). Adding the single clone function is cleaner/easier.

File: include/grpc/census.h (hunk 3 above)
Comment (a-veitch) [inline include/grpc/census.h]: Inside the structure, it works fine - I think it works OK. If it causes problems later, we can rename it without breaking the ABI.

File: include/grpc/census.h (hunk 4 above)
Comment (a-veitch) [inline include/grpc/census.h]: Changed to sum.

File: include/grpc/census.h [hunk 5]
```diff
@@ -379,16 +379,131 @@ int census_tag_set_next(census_tag_set_iterator *it, census_tag_const *tag);
...
   double value;
-} census_stat;
+} census_value;
+
+/* Record new usage values against the given context. */
+void census_record_values(census_context *context, census_value *values,
+                          size_t nvalues);
+
+/** Structure used to describe an aggregation type. */
+typedef struct {
```
Comment (ctiller) [inline include/grpc/census.h]: Could we make this an internal type? (under the src/core/... tree?)

File: src/core/census/rpc_metric_id.h [hunk 6]
```diff
@@ -31,16 +31,21 @@
...
-/* Stats ID's used for RPC measurements. */
-#define CENSUS_INVALID_STAT_ID 0     /* ID 0 is always invalid */
-#define CENSUS_RPC_CLIENT_REQUESTS 1 /* Count of client requests sent. */
-#define CENSUS_RPC_SERVER_REQUESTS 2 /* Count of server requests sent. */
-#define CENSUS_RPC_CLIENT_ERRORS 3   /* Client error counts. */
-#define CENSUS_RPC_SERVER_ERRORS 4   /* Server error counts. */
-#define CENSUS_RPC_CLIENT_LATENCY 5  /* Client side request latency. */
-#define CENSUS_RPC_SERVER_LATENCY 6  /* Server side request latency. */
+/* Metric ID's used for RPC measurements. */
+/* Count of client requests sent. */
```
Comment (maxwell-demon) [inline src/core/census/rpc_metric_id.h]: Consider keeping CENSUS_INVALID_METRIC_ID.

File: include/grpc/census.h (hunk 5 above)
Comment (a-veitch) [inline include/grpc/census.h]: We could. Advantage is that it declutters the API some; disadvantage is that we then wouldn't be able to support user-defined aggregation types. Thoughts/opinions on that tradeoff?

File: src/core/census/rpc_metric_id.h (hunk 6 above)
Comment (a-veitch) [inline src/core/census/rpc_metric_id.h]: > Consider keeping CENSUS_INVALID_METRIC_ID.

Why? What would we use it for?

Comment (maxwell-demon) [inline src/core/census/rpc_metric_id.h]: Just like the default/stale handle, one use case is to answer the query "how many record_stats() calls have invalid metric id across different processes?", we can use it as a measure of the quality of stats.

There are various scenarios that two processes (P_A and P_B) may see different metric id definitions (e.g. They are running different versions of binaries, or, there's a long latency to get updated metric ids from a centralized server). Having an common INVALID_METRIC_ID can allow us to globally collect and aggregate the stats of "record against invalid metric id".

Comment (maxwell-demon) [inline src/core/census/rpc_metric_id.h]: Really not a strong use case, just something I immediately thought about after seeing invalid resource id got removed. 

I don't insist on keeping it.  It's your call.

Comment (a-veitch) [inline src/core/census/rpc_metric_id.h]: Going to keep it as-is - think we can keep a count of invalid metric ID's seen, without having to reserve an ID entry for it.

File: include/grpc/census.h (hunk 5 above)
Comment (ctiller) [inline include/grpc/census.h]: Advantage is that we could tweak it as needed without backwards compatibility worries.

I'd prefer to be conservative for now, and maybe expand later with some more experience.

Comment (a-veitch) [inline include/grpc/census.h]: Done

**Pull request merged**